aioesphomeapi/connection.py

```diff
@@ -109,34 +109,40 @@ def __init__(
 
         self._process_task: Optional[asyncio.Task[None]] = None
 
+        self._connect_lock: asyncio.Lock = asyncio.Lock()
+
     async def _cleanup(self) -> None:
         """Clean up all resources that have been allocated.
 
         Safe to call multiple times.
         """
-        if self._frame_helper is not None:
-            await self._frame_helper.close()
-            self._frame_helper = None
-
-        if self._process_task is not None:
-            self._process_task.cancel()
-            with suppress(asyncio.CancelledError):
-                await self._process_task
-            self._process_task = None
-
-        if self._socket is not None:
-            self._socket.close()
-            self._socket = None
-
-        if not self._on_stop_called and self._connect_complete:
-            # Ensure on_stop is called
-            asyncio.create_task(self.on_stop())
-            self._on_stop_called = True
-
-        # Note: we don't explicitly cancel the ping/read task here
-        # That's because if not written right the ping/read task could cancel
-        # themself, effectively ending execution after _cleanup which may be unexpected
-        self._ping_stop_event.set()
+        async def _do_cleanup():
+            async with self._connect_lock:
+                if self._frame_helper is not None:
+                    await self._frame_helper.close()
+                    self._frame_helper = None
+
+                if self._process_task is not None:
+                    self._process_task.cancel()
+                    with suppress(asyncio.CancelledError):
+                        await self._process_task
+                    self._process_task = None
+
+                if self._socket is not None:
+                    self._socket.close()
+                    self._socket = None
+
+                if not self._on_stop_called and self._connect_complete:
+                    # Ensure on_stop is called
+                    asyncio.create_task(self.on_stop())
+                    self._on_stop_called = True
+
+                # Note: we don't explicitly cancel the ping/read task here
+                # That's because if not written right the ping/read task could cancel
+                # themself, effectively ending execution after _cleanup which may be unexpected
+                self._ping_stop_event.set()
+
+        asyncio.create_task(_do_cleanup())
```

Comment:
It looks like this task could end up still running while it's trying to connect again when it awaits and returns control to the event loop

Reply:
That's right, but the key is that it doesn't perform disconnect actions until after the initial connect has returned. Specifically, `on_stop` is called after preventing a permanent disconnect scenario.

Reply:
I think it would be a bit cleaner to move the contents of _do_cleanup into a new function called _do_cleanup_locked, hold the lock in _do_cleanup, and call _do_cleanup_locked under the lock and then you can call the _do_cleanup_locked function from the exception handler in the connect without releasing the lock or creating another task

Reply:
To be clear, is the main benefit of this to simply avoid creating a new task in the one scenario with an exception case in connect? Am I missing something or are there additional benefits to doing so? It seems that would make things a lot more complicated for seemingly little gain.

Reply:
I can look into storing the task and only queuing a new one if there is no existing task, but I would have to guess there's little performance to be gained from this.

Reply:
> To be clear, is the main benefit of this to simply avoid creating a new task in the one scenario with an exception case in connect? Am I missing something or are there additional benefits to doing so? It seems that would make things a lot more complicated for seemingly little gain.

I'm more concerned about future potential races that could happen between when control is returned to the event loop and then the lock is reacquired to execute the cleanup code. 

Since type of races are very difficult to track down so I always prefer a design where we hold the connection lock for the duration of the connection attempt and subsequent cleanup so we don't have analyze every connect and disconnect call to make sure it's race safe

Reply:
Does cherry-picking this commit https://github.com/esphome/aioesphomeapi/commit/79e520d9f6ed56a5a49726d6d96e06a737c633e2 still solve the issue you are seeing?

Reply:
That would result in a deadlock because `report_fatal_error` would get called which calls cleanup.

Reply:
For the purposes of answer the above question, please assume that would be refactored to not be a problem

Reply:
I'm not sure I understand. Unless there is a major refactor to the code, I'm not sure how it would be fixed to not result in a deadlock. If there is such a major refactor, I don't see the point of reviewing this change. The connect code relies on errors being raised and passed through. These errors are followed the cleanup method

Reply:
The one in `send_message_await_response_complex` that calls `send_message` is a bit unexpected since it will never await the future it creates because if `send_message` raise `SocketAPIError` which means the the `_read_exception_handlers` never get cleaned up.  I think it is not a problem in practice since I don't think we reuse the APIConnection object.

But this code path could definitely use some refactoring as its far too complex to follow

Reply:
I hate to add another flag, but I don't think you wanted to sign up for refactoring this.

Something like this will work and still be race safe https://github.com/esphome/aioesphomeapi/commit/5017794de616de5b6de2af963d5b29a1c5096d40

Reply:
I see another call to `cleanup` at line 569 that might also deadlock in this situation. The ultimate solution to this might be a larger refactor of the entire file. I can see wanting to avoid lingering tasks, but this likely can be avoided by having a single task that is cancelled during shutdown.